### ISSUE TITLE
Backport all 20 provable-contracts kernel contracts into aprender test suite

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#285]
## Summary

provable-contracts now has **20 validated kernel contracts** covering Tier 1 (core transformer), Tier 2 (compound), and Tier 3 (classical ML). aprender currently only wires **6 of 20** contracts into its test suite. This ticket tracks backporting the remaining 14 contracts as wired property tests.

### Already Wired (6/20)

| Contract | Test File | Status |
|---|---|---|
| `softmax-kernel-v1.yaml` | `tests/contracts/softmax_contract.rs` | Passing |
| `rmsnorm-kernel-v1.yaml` | `tests/contracts/rmsnorm_contract.rs` | Passing |
| `rope-kernel-v1.yaml` | `tests/contracts/rope_contract.rs` | Passing |
| `attention-kernel-v1.yaml` | `tests/contracts/attention_contract.rs` | Passing |
| `activation-kernel-v1.yaml` | `tests/contracts/activation_contract.rs` | Passing |
| `matmul-kernel-v1.yaml` | `tests/contracts/matmul_contract.rs` | Passing |

#### Tier 1 — Directly testable (implementation exists)

| Contract | aprender Function | Signature | File |
|---|---|---|---|
| `layernorm-kernel-v1.yaml` | `LayerNorm::forward` | `fn forward(&self, input: &Tensor) -> Tensor` | `src/nn/normalization/mod.rs:98` |
| `cross-entropy-kernel-v1.yaml` | `CrossEntropyLoss::forward` | `fn forward(&self, logits: &Tensor, targets: &Tensor) -> Tensor` | `src/nn/loss.rs` |
| `adamw-kernel-v1.yaml` | `AdamW::step` | `fn step(&mut self)` (Optimizer trait) | `src/nn/optim/mod.rs:384` |
| `batchnorm-kernel-v1.yaml` | `BatchNorm1d::forward` | `fn forward(&mut self, input: &Tensor) -> Tensor` | `src/nn/normalization/mod.rs:175` |
| `conv1d-kernel-v1.yaml` | `Conv1d::forward` | `fn forward(&self, input: &Tensor) -> Tensor` | `src/nn/conv/mod.rs:40` |
| `kmeans-kernel-v1.yaml` | `KMeans::fit` / `predict` | `fn fit(&mut self, data: &Matrix<f32>)` / `fn predict(&self, data: &Matrix<f32>) -> Vec<usize>` | `src/cluster/kmeans.rs:50` |
| `pagerank-kernel-v1.yaml` | `GraphCentrality::pagerank` | `fn pagerank(&self, damping: f64, max_iter: usize, tol: f64) -> Result<Vec<f64>>` | `src/graph/centrality.rs:64` |
| `lbfgs-kernel-v1.yaml` | `LBFGS::minimize` | `fn minimize<F, G>(&mut self, f: F, grad: G, x0: Vector<f32>) -> OptimizationResult<f32>` | `src/optim/lbfgs.rs:62` |
| `cma-es-kernel-v1.yaml` | `CmaEs::optimize` | `fn optimize<F>(&mut self, obj: F, space: &SearchSpace, budget: Budget) -> OptimizationResult<f64>` | `src/metaheuristics/cmaes.rs:54` |

#### Tier 2 — Partially testable or blocked

| Contract | Issue | Action |
|---|---|---|
| `swiglu-kernel-v1.yaml` | `swiglu_fused` is private in `models::qwen2` | Mark `#[ignore]` until public API exposed |
| `gqa-kernel-v1.yaml` | GQA logic embedded in `MultiHeadAttention` | Mark `#[ignore]` until separately testable |
| `silu-kernel-v1.yaml` | SiLU only exists as private fn | Mark `#[ignore]`, create sub-issue to add `nn::functional::silu` |
| `ssm-kernel-v1.yaml` | SSM/Mamba not implemented | Mark `#[ignore]` |
| `flash-attention-v1.yaml` | Flash attention not implemented | Mark `#[ignore]` |

### Step 1: Update Makefile CONTRACTS list

Add all 13 new contracts to the `CONTRACTS` variable:

```makefile
CONTRACTS := $(PV_DIR)/contracts/softmax-kernel-v1.yaml \
             $(PV_DIR)/contracts/rmsnorm-kernel-v1.yaml \
             $(PV_DIR)/contracts/rope-kernel-v1.yaml \
             $(PV_DIR)/contracts/attention-kernel-v1.yaml \
             $(PV_DIR)/contracts/activation-kernel-v1.yaml \
             $(PV_DIR)/contracts/matmul-kernel-v1.yaml \
             $(PV_DIR)/contracts/flash-attention-v1.yaml \
             $(PV_DIR)/contracts/swiglu-kernel-v1.yaml \
             $(PV_DIR)/contracts/gqa-kernel-v1.yaml \
             $(PV_DIR)/contracts/layernorm-kernel-v1.yaml \
             $(PV_DIR)/contracts/silu-kernel-v1.yaml \
             $(PV_DIR)/contracts/cross-entropy-kernel-v1.yaml \
             $(PV_DIR)/contracts/adamw-kernel-v1.yaml \
             $(PV_DIR)/contracts/ssm-kernel-v1.yaml \
             $(PV_DIR)/contracts/conv1d-kernel-v1.yaml \
             $(PV_DIR)/contracts/batchnorm-kernel-v1.yaml \
             $(PV_DIR)/contracts/kmeans-kernel-v1.yaml \
             $(PV_DIR)/contracts/pagerank-kernel-v1.yaml \
             $(PV_DIR)/contracts/lbfgs-kernel-v1.yaml \
             $(PV_DIR)/contracts/cma-es-kernel-v1.yaml
```

### Step 2: Write 14 wired test files

For each contract, create `tests/contracts/<kernel>_contract.rs` following the established pattern:

1. Run `pv probar <contract>.yaml --binding contracts/aprender/binding.yaml` to get the scaffold
2. Replace stub calls with real aprender function calls (signatures above)
3. Hand-tune proptest ranges for numerical stability (see `softmax_contract.rs` as reference)
4. Mark SIMD equivalence tests as `#[ignore = "SIMD equivalence — trueno domain"]`
5. Mark unimplemented features as `#[ignore = "<reason>"]`

**Files to create:**

```
tests/contracts/layernorm_contract.rs       # 6 obligations, 7 falsification tests
tests/contracts/cross_entropy_contract.rs   # 5 obligations, 6 falsification tests
tests/contracts/adamw_contract.rs           # 5 obligations, 6 falsification tests
tests/contracts/batchnorm_contract.rs       # 5 obligations, 6 falsification tests
tests/contracts/conv1d_contract.rs          # 5 obligations, 6 falsification tests
tests/contracts/kmeans_contract.rs          # 5 obligations, 6 falsification tests
tests/contracts/pagerank_contract.rs        # 5 obligations, 6 falsification tests
tests/contracts/lbfgs_contract.rs           # 5 obligations, 6 falsification tests
tests/contracts/cma_es_contract.rs          # 5 obligations, 6 falsification tests
tests/contracts/swiglu_contract.rs          # 4 obligations (all #[ignore] — private API)
tests/contracts/gqa_contract.rs             # 5 obligations (all #[ignore] — embedded)
tests/contracts/silu_contract.rs            # 5 obligations (all #[ignore] — private)
tests/contracts/ssm_contract.rs             # 5 obligations (all #[ignore] — not impl)
tests/contracts/flash_attention_contract.rs # 4 obligations (all #[ignore] — not impl)
```

### Step 3: Update `tests/contracts/mod.rs`

Add 14 new module declarations:

```rust
mod layernorm_contract;
mod cross_entropy_contract;
mod adamw_contract;
mod batchnorm_contract;
mod conv1d_contract;
mod kmeans_contract;
mod pagerank_contract;
mod lbfgs_contract;
mod cma_es_contract;
mod swiglu_contract;
mod gqa_contract;
mod silu_contract;
mod ssm_contract;
mod flash_attention_contract;
```

### Step 4: Verify

```bash
make contract-validate   # All 20 contracts valid
make contract-test       # All property tests pass
make contract-audit      # All bindings accounted for
cargo test --test contract_tests -- --include-ignored  # See ignored count
```

## Test Pattern Reference

Each wired test file follows this pattern (from `softmax_contract.rs`):

```rust
// CONTRACT: <kernel>-kernel-v1.yaml
// HASH: sha256:<hash>
// Generated by: pv probar --binding
// DO NOT EDIT — regenerate with `pv probar --binding`

use aprender::autograd::Tensor;
use aprender::nn::<Module>;
use proptest::prelude::*;

proptest! {
    /// Obligation: <property> (<type>)
    /// Formal: <mathematical statement>
    #[test]
    fn prop_<property_name>(
        data in proptest::collection::vec(-100.0f32..100.0, 1..64usize)
    ) {
        let n = data.len();
        let x = Tensor::new(&data, &[1, n]);
        let y = <real_aprender_call>(&x);
        // Assert obligation holds
        prop_assert!(<condition>, "<diagnostic message>");
    }
}
```

## Obligations Summary

| Tier | Contracts | Obligations | Falsification Tests | Kani Harnesses |
|---|---|---|---|---|
| Core (existing 7) | 7 | 35 | 32 | 11 |
| Tier 1 new (SwiGLU, GQA, LayerNorm) | 3 | 15 | 19 | 9 |
| Tier 2 new (SiLU, CE, AdamW) | 3 | 15 | 18 | 9 |
| Tier 3 (SSM, Conv1d, BN, KM, PR, LB, CMA) | 7 | 35 | 42 | 14 |
| **Total** | **20** | **101** | **115** | **46** |

## Definition of Done

- [ ] Makefile `CONTRACTS` variable lists all 20 contracts
- [ ] 14 new test files in `tests/contracts/`
- [ ] `tests/contracts/mod.rs` declares all 20 modules
- [ ] `make contract-validate` passes (20/20)
- [ ] `make contract-test` passes (all property tests green)
- [ ] `make contract-audit` passes (all bindings clean)
- [ ] 9 contracts fully wired with real function calls
- [ ] 5 contracts with `#[ignore]` stubs (blocked on API exposure)
- [ ] CI integration: contract tests run in `make test-full`

## Follow-up Issues

After this ticket, consider:
1. **Expose `nn::functional::silu`** — unblocks `silu-kernel-v1.yaml` and `swiglu-kernel-v1.yaml` wiring
2. **Extract GQA from MultiHeadAttention** — unblocks `gqa-kernel-v1.yaml` wiring
3. **Implement SSM/Mamba** — unblocks `ssm-kernel-v1.yaml` wiring
4. **Implement Flash Attention** — unblocks `flash-attention-v1.yaml` wiring
5. **Run Kani harnesses** — 46 bounded model checking proofs across all 20 contracts

## References

- provable-contracts registry: `paiml/provable-contracts` (20 contracts, 101 obligations)
- Binding registry: `contracts/aprender/binding.yaml` (32 binding entries)
- Kernel contracts spec: `docs/specifications/kernel-contracts.md`
- Existing pattern: `tests/contracts/softmax_contract.rs` (6 property tests, all passing)
